adapters/controllers/basic_controller.go

```diff
@@ -0,0 +1,61 @@
+package controllers
+
+import (
+	"encoding/json"
+	"net/http"
+	"wizeline/ghibli/services/repository"
+)
+
+type BasicController[T repository.Indexable] struct {
+	repo repository.Repository[T]
+}
+
+func NewBasicController[T repository.Indexable](
+	repo repository.Repository[T]) BasicController[T] {
+	return BasicController[T]{repo: repo}
+}
+
+func sendEncoded(w http.ResponseWriter, data []byte) {
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+	w.Write(data)
+}
+
+func (f BasicController[T]) getEncodedFilms() ([]byte, error) {
+	films, err := f.repo.ListAll()
+	if err != nil {
+		return []byte{}, err
+	}
+
+	return json.MarshalIndent(films, "", "\t")
+}
+
+func (f BasicController[T]) getEncodedFilm(id string) ([]byte, error) {
+	film, err := repository.GetById(f.repo, id)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	return json.MarshalIndent(film, "", "\t")
+}
+
+func (f BasicController[T]) tryGetEncoded(r *http.Request) ([]byte, error) {
+	id := r.URL.Query().Get("id")
+
+	switch {
+	case id != "":
+		return f.getEncodedFilm(id)
+	default:
+		return f.getEncodedFilms()
+	}
+}
+
+func (f BasicController[T]) ServeFilms(w http.ResponseWriter, r *http.Request) {
+	data, err := f.tryGetEncoded(r)
+
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusNotFound)
+	} else {
+		sendEncoded(w, data)
+	}
```

Comment:
You could skip the `else` sentence.

```suggestion
	if err != nil {
		http.Error(w, err.Error(), http.StatusNotFound)
	}

	sendEncoded(w, data)
```

Also, you might want to review what kind of errors you are sending.

Reply:
Thanks, this was corrected and rebased from the first deliverable.